chia/cmds/param_types.py

```
@@ -0,0 +1,166 @@
+        if self.address_type is not address_type:
+            raise ValueError(f"Address must be of type {address_type}")
+        return self.puzzle_hash
+
+
+class AddressParamType(click.ParamType):
+    """
+    A Click parameter type for bech32m encoded addresses, it gives a class with the address type and puzzle hash.
+    """
+
+    name: str = "CliAddress"  # output type
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> CliAddress:
+        if isinstance(value, CliAddress):  # required by click
+            return value
+        try:
+            hrp, b32data = bech32_decode(value)
+            if hrp in ["xch", "txch"]:  # I hate having to load the config here
+                expected_prefix = ctx.obj.get("expected_prefix") if ctx else None
+                if expected_prefix is None:
+                    from chia.util.config import load_config
+                    from chia.util.default_root import DEFAULT_ROOT_PATH
+
+                    root_path = ctx.obj["root_path"] if ctx is not None else DEFAULT_ROOT_PATH
+                    config = load_config(root_path, "config.yaml")
+                    expected_prefix = selected_network_address_prefix(config)
+                    if ctx is not None:
+                        ctx.obj["expected_prefix"] = expected_prefix
+                # now that we have the expected prefix, we can validate the address
+                if hrp == expected_prefix:
+                    addr_type = AddressType.XCH
+                else:
+                    self.fail(f"Unexpected Address Prefix: {hrp}, are you sure its for the right network?", param, ctx)
+            else:
+                addr_type = AddressType(hrp)
+            return CliAddress(puzzle_hash=decode_puzzle_hash(value), address_type=addr_type, original_address=value)
+        except ValueError:
+            self.fail("Address must be a valid bech32m address", param, ctx)
```

don't you need to re-throw the exception here?

---

`ParamType.fail` is what the docs said to use for an invalid value message: [https://click.palletsprojects.com/en/8.0.x/api/?highlight=type#click.ParamType.fail](https://click.palletsprojects.com/en/8.0.x/api/?highlight=type#click.ParamType.fail)

---

Should we consider (here and below and wherever) to include the failure?  The user getting a message pointing out a length, or invalid character, etc would help them identify their error.